contracts/near/eth2-client/src/lib.rs

```diff
@@ -259,41 +214,86 @@ impl Eth2Client {
 
     #[result_serializer(borsh)]
     pub fn submit_execution_header(&mut self, #[serializer(borsh)] block_header: BlockHeader) {
-        if self.finalized_beacon_header.execution_block_hash != block_header.parent_hash {
-            self.unfinalized_headers
-                .get(&block_header.parent_hash)
-                .unwrap_or_else(|| {
-                    env::panic_str(
-                        format!(
-                            "Header has unknown parent {:?}. Parent should be submitted first.",
-                            block_header.parent_hash
-                        )
-                        .as_str(),
-                    )
-                });
-        }
+        require!(self.client_mode == ClientMode::SubmitHeader);
 
-        let submitter = env::predecessor_account_id();
-        self.update_submitter(&submitter, 1);
         let block_hash = block_header.calculate_hash();
-        #[cfg(feature = "logs")]
-        env::log_str(
+        let expected_block_hash = self
+            .unfinalized_tail_execution_header
+            .as_ref()
+            .map(|header| header.parent_hash)
+            .unwrap_or(self.finalized_beacon_header.execution_block_hash);
+        require!(
+            block_hash == expected_block_hash,
             format!(
-                "Submitted header number {}, hash {:?}",
-                block_header.number, block_hash
+                "The expected block hash is {:#?} but got {:#?}.",
+                expected_block_hash, block_hash
```

Comment:
Maybe we can also print the block number and expected block number? I think it can be useful in the future for debug.

Reply:
Btw, I've improved the logs a little bit here 
abe2d7b09e5be568d7463210002ed9d823d986d8